lib/assertions.js

```diff
@@ -892,7 +895,177 @@ module.exports = function (expect) {
         expect(subject, 'to equal', value);
     });
 
-    expect.addAssertion('<object> to [exhaustively] satisfy <array-like|object>', function (expect, subject, value) {
+    expect.addAssertion('<array-like> to [exhaustively] satisfy <array-like>', function (expect, subject, value) {
+        expect.errorMode = 'bubble';
+        var keyPromises = new Array(value.length);
+        var i;
+        var valueKeys = new Array(value.length);
+        for (i = 0 ; i < value.length ; i += 1) {
+            valueKeys[i] = i;
+            keyPromises[i] = expect.promise(function () {
+                var valueKeyType = expect.findTypeOf(value[i]);
+                if (valueKeyType.is('function')) {
+                    return value[i](subject[i]);
+                } else {
+                    return expect(subject[i], 'to [exhaustively] satisfy', value[i]);
+                }
+            });
+        }
+        return expect.promise.all([
+            expect.promise(function () {
+                expect(subject, 'to only have keys', valueKeys);
+            }),
+            expect.promise.all(keyPromises)
+        ]).caught(function () {
+            var subjectType = expect.subjectType;
+            return expect.promise.settle(keyPromises).then(function () {
+                var toSatisfyMatrix = new Array(subject.length * value.length);
+                for (i = 0 ; i < subject.length ; i += 1) {
+                    if (i < value.length) {
+                        toSatisfyMatrix[i + i * subject.length] = keyPromises[i].isFulfilled() || keyPromises[i].reason();
+                    }
+                }
+                if (subject.length > 10 || value.length > 10) {
```

Comment:
Ah, the `toSatisfyMatrix` is used to build the diff in the index-by-index case as well: https://github.com/unexpectedjs/unexpected/pull/231/files#diff-0a56a2b979f6080486ec1909508cda1aR1037